pkg/pki/signer/signer.go

```diff
@@ -45,17 +46,37 @@ func NewSigner(caFile, caKeyFile string, duration time.Duration) (*Signer, error
 	return ret, nil
 }
 
-func (s *Signer) Sign(x509cr *x509.CertificateRequest, usages []capi.KeyUsage) ([]byte, error) {
+func (s *Signer) Sign(x509cr *x509.CertificateRequest, spec capi.CertificateSigningRequestSpec) ([]byte, error) {
 	currCA, err := s.caProvider.currentCA()
 	if err != nil {
 		return nil, err
 	}
 	der, err := currCA.Sign(x509cr.Raw, authority.PermissiveSigningPolicy{
-		TTL:    s.certTTL,
-		Usages: usages,
+		TTL:    s.duration(spec.ExpirationSeconds),
+		Usages: spec.Usages,
 	})
 	if err != nil {
 		return nil, err
 	}
 	return pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: der}), nil
 }
+
+func (s *Signer) duration(expirationSeconds *int32) time.Duration {
+	if expirationSeconds == nil {
+		return s.certTTL
+	}
+
+	// honor requested duration is if it is less than the default TTL
+	// use 10 min (2x hard coded backdate above) as a sanity check lower bound
+	const min = 10 * time.Minute
+	switch requestedDuration := csr.ExpirationSecondsToDuration(*expirationSeconds); {
+	case requestedDuration > s.certTTL:
```

Comment:
Thank you for the contribution @enj.

I have a question: why not always uses the requestedDuration even it's greater than certificate TTL?

Reply:
I followed the semantics used by the [KCM signer](https://github.com/kubernetes/kubernetes/blob/4d08582d1fa21e1f5887e73380001ac827371553/pkg/controller/certificates/signer/signer.go#L209-L231) which treats `certTTL` as the max duration the signer will accept (i.e. it is a policy decision to limit the caller from asking for a certificate with a very long lifetime).  I think an admin would find it surprising if a user could ask for a certificate with a longer lifespan than `certTTL`.